### ISSUE TITLE
Add support for generating resource-config.json file

### ISSUE THREAD

**Issue by melix** [graalvm/native-build-tools#75]
## Proposal

Java applications often need resources which are found on classpath. While those resources are visible at native image build time, they are not included at runtime, unless you provide a `resource-config.json` file. In the Java ecosystem, and in particular in the Maven and Gradle build systems, resources are often found in a conventional place (`src/main/resources`) or generated in a well-known location.

This means that we could provide a feature to _infer_ the resources to be included in the image, and therefore generate the `resources-config.json` file automatically.

Ideally, this option should be opt-in (or opt-out). Things to consider:

- multiple source sets (`src/main/resources`, `src/feature/resources`, ...)
- resources from transitive dependencies (projects/modules and external dependencies)
- potential filtering

See https://github.com/micronaut-projects/micronaut-gradle-plugin/pull/215 for a project which works around this limitation.

**Comment by sdeleuze** [comment]
As discussed, I would limit the defaults to something very generic and not opinionated (`src/main/resources`, `src/test/resources`).

**Comment by melix** [comment]
Please take a look at the commit mentioned above. It introduces resource config file generation. By default, it is very conservative, with no inference and only explicit configuration. However, it supports inference for project dependencies and transitive dependencies (optionally).

**Comment by sdeleuze** [comment]
Thanks for sharing.

Could we define what is inference behavior by default? For project dependencies, is it only using `src/main/resources` and `src/test/resources` (or similar customized resource directories) content for the various modules?

For external dependencies, is it everything except `*.class`?

Do we agree that we need to provide a similar capability in Maven side?

**Comment by melix** [comment]
So the inference process is consistent and doesn't make a difference between project dependencies and external dependencies (but inference for external dependencies is turned off by default). The source set is the wrong abstraction level. Basically when you build a native image, what you include is a number of classes and resources which come from a number of source sets, not necessarily one. It's a common mistake to have Gradle plugins which rely on specific source sets, because they typically break as soon as the user configures additional source sets to be included in the application.

Therefore, I think the right way to implement inference is what I have done in this PR, which is to analyze the classpath, scan every entry, and decide what do to:

- exclude all `*.class` files which are obviously not resources
- exclude cases where a `META-INF/native-image` folder is present, in which case we know we don't have to infer anything
- exclude a number of common patterns by default (but again overridable)

Basically it means that you'd get the resources from `src/main/resources` _and_ resources which are included in the app because extra source sets are included (say, `src/myfeature/resources`).

An exception to the rule is made for _tests_, in which case not only do we add the usual application classpath, but we, this time, include the resources from the _test source set_, because, in this case, we have a _test source set_ (in this case it makes sense).

**Comment by sdeleuze** [comment]
Make sense.

I am wondering if a part of that should go at `native-image` level, especially given the fact most of it do not depend on source sets. I would like to avoid too much duplication of such basic capabilities across each build system we support. cc @vjovanov

**Comment by graemerocher** [comment]
Will this approach work for `nativeRun`? ie does `nativeRun` build the JAR and put that on the classpath or does it put the project build classes/resources directories on the classpath?

**Comment by melix** [comment]
Currently the plugin uses the jars, that is to say that to build a native image it first builds the application jar and then uses it.
